server/core/include/irods/irods_notify_service_manager.hpp

```diff
@@ -0,0 +1,56 @@
+#ifndef IRODS_NOTIFY_SERVICE_MANAGER_HPP
+#define IRODS_NOTIFY_SERVICE_MANAGER_HPP
+
+#include <fmt/format.h>
+#include <fmt/compile.h>
+
+#include <cstddef>
+#include <ctime>
+#include <string>
+#include <utility>
+
+namespace irods
+{
+    void notify_service_manager(const std::size_t _msg_length, const char* _msg, const char* _socket_path);
+
+    static inline void notify_service_manager(const std::string& _msg)
+    {
+        const char* sm_socket_path = std::getenv("NOTIFY_SOCKET");
+        if (sm_socket_path == nullptr) {
+            // if NOTIFY_SOCKET is not set, we're done
+            return;
+        }
+
+        notify_service_manager(_msg.size(), _msg.data(), sm_socket_path);
+    }
+
+    template <typename... Args>
+    static inline void notify_service_manager(const fmt::format_string<Args...>& _format, Args&&... _args)
+    {
+        const char* sm_socket_path = std::getenv("NOTIFY_SOCKET");
+        if (sm_socket_path == nullptr) {
+            // if NOTIFY_SOCKET is not set, we're done
+            return;
+        }
+
+        auto msg = fmt::format(_format, std::forward<Args>(_args)...);
+        notify_service_manager(msg.size(), msg.data(), sm_socket_path);
+    }
+
+    template <typename CompiledFormat, typename... Args>
+        requires fmt::detail::is_compiled_string<CompiledFormat>::value ||
+                 fmt::detail::is_compiled_format<CompiledFormat>::value
+    static inline void notify_service_manager(const CompiledFormat& _format, Args&&... _args)
```

Comment:
If the components within the detail namespace change in between releases, then we may have to revisit this code which I'd prefer no one ever have to do.

What if we just didn't use a compiled fmt string?
Is there a significant reason to use them for the service mgr?

Reply:
Compiled format strings offer a performance benefit, though I'm not sure how much.

Given that we'd like to be able to also accept compiled format strings in the logger (#8101) (which would potentially offer a much larger performance boon), I'll open an issue about this in the upstream fmt repo.

Reply:
That's true, but they also have some downsides. See https://fmt.dev/11.1/api/#format-string-compilation. I'm pretty sure the 8.1 docs state a similar thing.
> Format string compilation can generate more binary code compared to the default API and is only recommended in places where formatting is a performance bottleneck.

Ignoring the compile time and code bloat, the performance bottleneck is the primary thing that determines whether we should use FMT_COMPILE. Until it's demonstrated that performance is an issue for the service manager implementation, I think we avoid use of FMT_COMPILE. I suspect the normal calls to fmtlib are good enough for our needs.

With that said, don't delete your implementation completely. Keep it in a branch until the fmtlib devs respond to your issue.

Reply:
I've created the branch `8036-fmt-compiled.main` in my fork and opened fmtlib/fmt#4335.